### PR TITLE
Add EC support for Z490 ROG MAXIMUS XII HERO (WI-FI)

### PR DESCRIPTION
**Pull request by Shadotty**
Added Intel400 series of motherboards and added the support for the mobo in the title. Already tested it both in LibreHardwareMonitor itself and also using the new .dll inside FanControl and everything works as intended.

Immagine 2024-06-05 235410
image

### CODE REVIEW

File: LibreHardwareMonitorLib/Hardware/Motherboard/Lpc/EC/EmbeddedController.cs
Comment (PhyxionNL) [inline LibreHardwareMonitorLib/Hardware/Motherboard/Lpc/EC/EmbeddedController.cs:324]: Have you checked?

Comment (Shadotty) [reply LibreHardwareMonitorLib/Hardware/Motherboard/Lpc/EC/EmbeddedController.cs:324]: Unfortunately I have no idea why it happens, the value at idle is correct but whenever I put a relatively important load on the CPU it becomes negative (and is of course completely wrong), so I don't know, should I make a new pr without this line of code?

Comment (Shadotty) [reply LibreHardwareMonitorLib/Hardware/Motherboard/Lpc/EC/EmbeddedController.cs:324]: @PhyxionNL I'm awaiting your reply, thanks

Comment (PhyxionNL) [reply LibreHardwareMonitorLib/Hardware/Motherboard/Lpc/EC/EmbeddedController.cs:324]: You can slash the broken line out first. Perhaps @zeule knows what's happening?

Comment (zeule) [reply LibreHardwareMonitorLib/Hardware/Motherboard/Lpc/EC/EmbeddedController.cs:324]: The value is unsigned, we read it a signed. Thanks for spotting the bug!

Comment (Shadotty) [reply LibreHardwareMonitorLib/Hardware/Motherboard/Lpc/EC/EmbeddedController.cs]: @zeule how do I fix it? for now i've just removed the line, I await instructions from you

Comment (zeule) [reply LibreHardwareMonitorLib/Hardware/Motherboard/Lpc/EC/EmbeddedController.cs]: @Shadotty before making changes, could you, please, confirm that that's indeed that case by observing at what value it flips? It should be at 128 if that's caused by the signed/unsigned mismatch.

Comment (Shadotty) [reply LibreHardwareMonitorLib/Hardware/Motherboard/Lpc/EC/EmbeddedController.cs]: I will check as soon as I can, it's difficult to get a 128A load but i will try

Comment (Shadotty) [reply LibreHardwareMonitorLib/Hardware/Motherboard/Lpc/EC/EmbeddedController.cs]: @zeule yes, I can confirm that the value becomes negative after exceeding 128A, so, what should be done here to fix this?

Comment (zeule) [reply LibreHardwareMonitorLib/Hardware/Motherboard/Lpc/EC/EmbeddedController.cs]: Thank you for testing that! I guess we can safely assume that this property is the same for all models, and them something like this should do the trick:
```diff
diff --git a/LibreHardwareMonitorLib/Hardware/Motherboard/Lpc/EC/EmbeddedController.cs b/LibreHardwareMonitorLib/Hardware/Motherboard/Lpc/EC/EmbeddedController.cs
index dc3399a..0e9b111 100644
--- a/LibreHardwareMonitorLib/Hardware/Motherboard/Lpc/EC/EmbeddedController.cs
+++ b/LibreHardwareMonitorLib/Hardware/Motherboard/Lpc/EC/EmbeddedController.cs
@@ -396,7 +396,7 @@ public abstract class EmbeddedController : Hardware
         {
             int val = _sources[si].Size switch
             {
-                1 => unchecked((sbyte)_data[readRegister]),
+                1 => _sources[si].Type switch { SensorType.Temperature => unchecked((sbyte)_data[readRegister]), _ => _data[readRegister] },
                 2 => unchecked((short)((_data[readRegister] << 8) + _data[readRegister + 1])),
                 _ => 0
             };
```

Comment (zeule) [reply LibreHardwareMonitorLib/Hardware/Motherboard/Lpc/EC/EmbeddedController.cs]: Turns out the Linux driver uses exactly this strategy for two years already (https://github.com/torvalds/linux/commit/8aba9ca62677570abf8c6b62611adb85bf1580ca), discovered in the same way (through CPU current), and I forgot to bring the change here. My bad…

Comment (Shadotty) [reply LibreHardwareMonitorLib/Hardware/Motherboard/Lpc/EC/EmbeddedController.cs]: So can I go ahead and add the line back to the code?

Comment (zeule) [reply LibreHardwareMonitorLib/Hardware/Motherboard/Lpc/EC/EmbeddedController.cs]: But of course.

Comment (Shadotty) [reply LibreHardwareMonitorLib/Hardware/Motherboard/Lpc/EC/EmbeddedController.cs:324]: Done, I'll wait for approval!

File: LibreHardwareMonitorLib/Hardware/Motherboard/Lpc/EC/EmbeddedController.cs
Comment (Shadotty) [reply LibreHardwareMonitorLib/Hardware/Motherboard/Lpc/EC/EmbeddedController.cs]: Sorry! I forgot it, fixed.

File: LibreHardwareMonitorLib/Hardware/Motherboard/Lpc/EC/EmbeddedController.cs
Comment (zeule) [inline LibreHardwareMonitorLib/Hardware/Motherboard/Lpc/EC/EmbeddedController.cs:322]: Please sort this by sensor address.

**Review by zeule — CHANGES_REQUESTED** [review]

File: LibreHardwareMonitorLib/Hardware/Motherboard/Lpc/EC/EmbeddedController.cs
Comment (Shadotty) [reply LibreHardwareMonitorLib/Hardware/Motherboard/Lpc/EC/EmbeddedController.cs]: Done, is this correct?

**Review by zeule — APPROVED** [review]
LGTM, thank you! Would appreciate if you share DSDT dump here or at https://github.com/zeule/asus-ec-sensors to include this board in the Linux HWMON driver. OR, perhaps, you'd like to submit the changes to the Linux kernel yourself?

**Review by zeule — APPROVED** [review]

File: LibreHardwareMonitorLib/Hardware/Motherboard/Lpc/EC/EmbeddedController.cs
Comment (zeule) [reply LibreHardwareMonitorLib/Hardware/Motherboard/Lpc/EC/EmbeddedController.cs]: Thanks!

**Comment by Shadotty** [comment]
> LGTM, thank you! Would appreciate if you share DSDT dump here or at https://github.com/zeule/asus-ec-sensors to include this board in the Linux HWMON driver. OR, perhaps, you'd like to submit the changes to the Linux kernel yourself?

Thank you!

I'm not going to lie this stuff is quite difficult for me to understand but I will for sure try to do it as soon as I can.

Just to be sure, now I just need to wait for someone to merge this PR right? Does it usually take long? I apologize but this is actually the first time that I contribute to something on github so this world is quite new for me.

**Comment by zeule** [comment]
> I'm not going to lie this stuff is quite difficult for me to understand but I will for sure try to do it as soon as I can.

To get the DSDT dump simply google "DSDT dump windows" and follow one of the many step-by-step guides. If this is your first GitHub experience, I'd rather suggest you to not contribute to the Linux project directly, as their bar is quite high.

> Just to be sure, now I just need to wait for someone to merge this PR right? Does it usually take long? I apologize but this is actually the first time that I contribute to something on github so this world is quite new for me.

@PhyxionNL  will merge the PR, should not take long.

**Comment by PhyxionNL** [comment]
Merged, thanks @zeule and @Shadotty. A follow-up PR to fix this bug would be appreciated as I cannot test iy myself :)

**Comment by zeule** [comment]
> Merged, thanks @zeule and @Shadotty. A follow-up PR to fix this bug would be appreciated as I cannot test iy myself :)

#1382, please.

**Pull request merged**
